### PR TITLE
[FEATURE] Add password strength validation (#9016)

### LINKED ISSUE #9016: [FEATURE] Add password strength validation with real-time feedback
As a **user**, I want **real-time password validation feedback** so that [**I can see which password requirements I've met while typing**].

## Context

Currently, omegaUp only validates password length (8-72 characters) on the backend via `SecurityTools::testStrongPassword()`. However:

1. **No complexity requirements**: Weak passwords like `12345678` or `password` are accepted
2. **No frontend validation**: Users only see errors after form submission
3. **Poor user experience**: Users don't know what requirements they need to meet

This creates a security vulnerability where user accounts can be easily compromised through brute-force or dictionary attacks.

### Backend Enhancement
Enhance [testStrongPassword()](cci:1://file:///c:/Users/BIT/Desktop/omegaup/frontend/server/src/SecurityTools.php:63:4-71:5) in [SecurityTools.php](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/server/src/SecurityTools.php:0:0-0:0) to require:
- At least 8 characters (already exists)
- At least one uppercase letter (A-Z)
- At least one lowercase letter (a-z)
- At least one digit (0-9)
- At least one special character (!@#$%^&*(),.?":{}|<>)

### Frontend Enhancement
Add real-time password validation feedback showing:

  Password requirements:
  - ✅ At least 8 characters
  - ❌ At least one uppercase letter (A-Z)
  - ✅ At least one lowercase letter (a-z)
  - ❌ At least one digit (0-9)
  - ❌ At least one special character (!@#$%^&*)

This feedback should appear on:
- User registration page
- Password change/reset forms
- Identity creation forms

## Benefits
- Improved account security
- Better user experience with clear, real-time feedback
- Alignment with industry security standards

### PR DESCRIPTION
**Pull request by Yashagarwal9798**
# Description

This PR implements password strength validation on both the backend and frontend, addressing issue #9016.

**Backend:**
- Enhanced `SecurityTools::testStrongPassword()` with regex validations requiring uppercase, lowercase, digit, and special character in passwords
- Added [SecurityToolsTest.php](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/tests/controllers/SecurityToolsTest.php:0:0-0:0) with 13 unit tests covering valid passwords, each rejection type, boundary cases, and common weak passwords

**Frontend:**
- Created [PasswordStrengthFeedback.vue](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.vue:0:0-0:0) — a reusable component that shows real-time ✅/❌ feedback for each password requirement as the user types
- Integrated the component into all 5 password forms: [Signup.vue](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/www/js/omegaup/components/login/Signup.vue:0:0-0:0), `PasswordEdit.vue`, `PasswordReset.vue`, `ChangePassword.vue`, and `PasswordAdd.vue`
- Added [PasswordStrengthFeedback.test.ts](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.test.ts:0:0-0:0) with 11 unit tests for the component

**Translations:**
- Added 9 new translation keys across all 4 languages (en, es, pt, pseudo) in [.ts](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/www/js/omegaup/lang.ts:0:0-0:0), [.lang](cci:7://file:///c:/Users/BIT/Desktop/omegaup/frontend/templates/en.lang:0:0-0:0), and [.json](cci:7://file:///c:/Users/BIT/Desktop/omegaup/package.json:0:0-0:0) file formats

Fixes: #9016

# Comments

The frontend regex patterns exactly match the backend PHP patterns to ensure consistency. The [PasswordStrengthFeedback](cci:2://file:///c:/Users/BIT/Desktop/omegaup/frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.vue:42:0-86:1) component uses Vue computed properties for reactive validation and emits a `validity-change` event that parent components can use to control form submission.

# Checklist:

- [x] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/blob/main/frontend/www/docs/Coding-guidelines.md) of omegaUp.
- [x] The tests were executed and all of them passed.
- [x] If you are creating a feature, the new tests were added.
- [ ] If the change is large (> 200 lines), this PR was split into various Pull Requests. It's preferred to create one PR for changes in controllers + unit tests in PHPUnit, and then another Pull Request for UI + tests in Jest, Cypress or both.

https://github.com/user-attachments/assets/2f9a955e-6b7c-40d1-9cfd-5db80f7927a1

### CODE REVIEW

**Comment by Yashagarwal9798** [comment]
@pabo99 sir , could you please review this PR? I have introduced several new test cases and modified existing ones, so I would appreciate your guidance to ensure the changes align with the project standards.

File: frontend/tests/controllers/SecurityToolsTest.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    public function testPasswordWithAllRequirementsPasses(): void {
+        // Multiple valid passwords
+        \OmegaUp\SecurityTools::testStrongPassword('Abc123!@#');
+        \OmegaUp\SecurityTools::testStrongPassword('Password1!');
+        \OmegaUp\SecurityTools::testStrongPassword('Test@123');
+        \OmegaUp\SecurityTools::testStrongPassword('MyP@ssw0rd');
+        $this->assertTrue(true);
+    }
+
+    /**
```
Comment (pabo99) [inline frontend/tests/controllers/SecurityToolsTest.php:28]: Could you add a parameterized test to validate all the following cases?
- `testPasswordTooShortRejected`
- `testPasswordTooLongRejected`
- `testPasswordWithoutUppercaseRejected`
- `testPasswordWithoutLowercaseRejected`
- `testPasswordWithoutDigitRejected`
- `testPasswordWithoutSpecialCharRejected`
- `testNullPasswordRejected`
- `testEmptyPasswordRejected`

This would reduce the size of the test while still covering all the scenarios

File: frontend/tests/controllers/SecurityToolsTest.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        // Test 'PASSWORD' - has uppercase but missing lowercase, digit, special
+        try {
+            \OmegaUp\SecurityTools::testStrongPassword('PASSWORD');
+            $this->fail('Expected exception for password "PASSWORD"');
+        } catch (\OmegaUp\Exceptions\InvalidParameterException $e) {
+            $this->assertSame('passwordMustContainLowercase', $e->getMessage());
+        }
+    }
```
Comment (pabo99) [inline frontend/tests/controllers/SecurityToolsTest.php]: It looks like these tests are duplicates of some of the previous ones. If you think it makes sense, you can include these cases in the parameterized test I suggested earlier

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -411,7 +411,7 @@ jobs:
       - name: Run the teaching assistant for course mode
         run: |
           cd stuff/teaching_assistant
-          python teaching_assistant.py --test_mode --skip-confirm --username teacher --password teacher123 --student_name student --language English --course_alias course --ta_feedback_indicator AI-generated --submission_id_mode false --assignment_alias assignment --llm omegaup --key omegaup
+          python teaching_assistant.py --test_mode --skip-confirm --username teacher --password 'Teacher_123!' --student_name student --language English --course_alias course --ta_feedback_indicator AI-generated --submission_id_mode false --assignment_alias assignment --llm omegaup --key omegaup
```
Comment (pabo99) [inline .github/workflows/ci.yml]: Oh, it's pretty bad that we're passing passwords in the CI workflows. The strangest part is that GitGuardian hasn't complained. Anyway, for now we'll leave it as is while we look for a solution to this

File: cypress/e2e/basic_commands.cy.ts [hunk 4]
```diff
@@ -36,7 +36,7 @@ describe('Basic Commands Test', () => {
   it('Should register a user using the API', () => {
     const loginOptions: LoginOptions = {
       username: uuid(),
-      password: uuid(),
+      password: 'P@55w0rd',
```
Comment (pabo99) [inline cypress/e2e/basic_commands.cy.ts]: I see you’re using this password in more than a couple of places in the Cypress tests. Could you add a constant with that password and reference it wherever it's needed? That way, in future changes we'd only need to modify it in one place

File: frontend/badges/500score/test.json [hunk 5]
```diff
@@ -6,7 +6,7 @@
       "apicalls": [
         {
           "username": "admintest",
-          "password": "testtesttest",
+          "password": "T3stt3st!",
```
Comment (pabo99) [inline frontend/badges/500score/test.json:9]: Same case for the passwords in the Jest tests. I see that here you might be using more than one different password. Perhaps you could create a pool of passwords and use whichever fits best for each test. 

What do you think?

File: frontend/tests/Utils.php [hunk 6]
```diff
@@ -161,7 +165,7 @@ private static function setUpDefaultDataConfig(): void {
         ['user' => $admin] = \OmegaUp\Test\Factories\User::createUser(
             new \OmegaUp\Test\Factories\UserParams([
                 'username' => 'admintest',
-                'password' => 'testtesttest',
+                'password' => 'T3stt3st!',
```
Comment (pabo99) [inline frontend/tests/Utils.php]: Just for consistency with the other tests, could you add this as a constant and use it in both places where it's needed?

File: frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.test.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
+import { shallowMount } from '@vue/test-utils';
+import PasswordStrengthFeedback from './PasswordStrengthFeedback.vue';
+
+describe('PasswordStrengthFeedback.vue', () => {
+  it('shows all requirements as unmet for empty password', () => {
+    const wrapper = shallowMount(PasswordStrengthFeedback, {
+      propsData: { password: '' },
+    });
+    const vm = wrapper.vm as any;
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.test.ts]: Could you avoid using the `any` type as much as possible?

It seems that in this context it's not necessary to explicitly indicate the type. Even better, you could use `wrapper.vm` directly

File: frontend/www/js/omegaup/components/user/PasswordAdd.vue [hunk 8]
```diff
@@ -57,10 +58,12 @@
 import { Vue, Component, Prop } from 'vue-property-decorator';
 import T from '../../lang';
 import omegaup_PasswordInput from '../common/PasswordInput.vue';
+import omegaup_PasswordStrengthFeedback from '../common/PasswordStrengthFeedback.vue';
 
 @Component({
   components: {
     'omegaup-password-input': omegaup_PasswordInput,
+    'omegaup-password-strength-feedback': omegaup_PasswordStrengthFeedback,
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/PasswordAdd.vue]: Is there a way for this component to be a direct child of `omegaup-password-input`? It seems the two will always be used together, so this would reduce usage to just one

File: frontend/www/js/omegaup/components/user/PasswordAdd.vue [hunk 9]
```diff
@@ -22,6 +22,7 @@
           :required="true"
           autocomplete="new-password"
         />
+        <omegaup-password-strength-feedback :password="newPassword" />
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/PasswordAdd.vue]: It seems a bit intrusive to have all that text showing:

Image

Is there a way to make it appear only once the password field loses focus? If that's the common behavior across different sites, we can leave it that way

File: frontend/www/js/omegaup/groups.ts [hunk 10]
```diff
@@ -102,28 +102,38 @@ export function getCSVRecords<T extends object>({
...
-    window.crypto.getRandomValues(arr);
-    return Array.from(
-      arr.filter((value) => value <= 255 - (255 % validChars.length)),
-      (value) => validChars[value % validChars.length],
-    )
-      .join('')
-      .substr(0, len);
+
+  const pickRandom = (chars: string): string =>
+    chars.charAt(Math.floor(Math.random() * chars.length));
```
Comment (pabo99) [inline frontend/www/js/omegaup/groups.ts]: Could you modify this function so that it uses `window.crypto.getRandomValues` when available, instead of `Math.random()`? This would make the password generation cryptographically secure. 

`Math.random()` could be used only as a fallback if `window.crypto` is not available

**Comment by Yashagarwal9798** [comment]
@pabo99 sir I have addressed all the changes you requested. I also took the opportunity to update the initial design.

Here is how it looks now:

https://github.com/user-attachments/assets/1d9a0a35-c8d8-4a8b-bda3-bedb725162c5

Let me know if you have any further feedback!

**Comment by Yashagarwal9798** [comment]
@pabo99 sir can you pls review this pr when you have time.

File: frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.vue [hunk 11]
```diff
@@ -0,0 +1,172 @@
+<template>
+  <div class="password-strength-feedback">
+    <div
+      :class="['strength-badge', isValid ? 'strong' : 'weak']"
+      @mouseenter="showDetails = true"
+      @mouseleave="showDetails = false"
+    >
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/common/PasswordStrengthFeedback.vue:7]: On mobile/touch devices, this is completely inaccessible, users will never see the requirements. The details should be shown by default (or at least when the field has content), not only on hover. Could you apply the changes?

File: frontend/tests/controllers/SecurityToolsTest.php [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+class SecurityToolsTest extends \OmegaUp\Test\ControllerTestCase {
+    /**
+     * Test that a strong password passes validation
+     */
+    public function testStrongPasswordPasses(): void {
+        // This should not throw any exception
+        \OmegaUp\SecurityTools::testStrongPassword('StrongP@ss1');
+        $this->assertTrue(true); // If we get here, the test passed
```
Comment (pabo99) [inline frontend/tests/controllers/SecurityToolsTest.php:13]: I don't really like this ant-pattern. It seems that `PHPUnit` has the correct method to handle this:

```suggestion
        $this->expectNotToPerformAssertions();
```

Same applies in the other places you use this pattern

**Review by pabo99 — COMMENTED** [review]
Two minor suggestions and it's okay for me!
